src/unix/fs.c

```diff
@@ -2003,14 +2003,19 @@ int uv_fs_read(uv_loop_t* loop, uv_fs_t* req,
   req->file = file;
 
   req->nbufs = nbufs;
-  req->bufs = req->bufsml;
-  if (nbufs > ARRAY_SIZE(req->bufsml))
-    req->bufs = uv__malloc(nbufs * sizeof(*bufs));
 
-  if (req->bufs == NULL)
-    return UV_ENOMEM;
+  if (cb != NULL) {
+    req->bufs = req->bufsml;
+    if (nbufs > ARRAY_SIZE(req->bufsml))
+      req->bufs = uv__malloc(nbufs * sizeof(*bufs));
+
+    if (req->bufs == NULL)
+      return UV_ENOMEM;
 
-  memcpy(req->bufs, bufs, nbufs * sizeof(*bufs));
+    memcpy(req->bufs, bufs, nbufs * sizeof(*bufs));
+  } else {
+    req->bufs = (uv_buf_t*)bufs; // Use bufs directly
```

Comment:
This is unsound because `uv_fs_req_cleanup()` will try to free() the memory. Set `req->bufs = NULL` after the operation completes.

Also, use C89 comments, not C99.